platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/BuildFailureProblemsCrossVersionSpec.groovy

```diff
@@ -0,0 +1,182 @@
+import org.gradle.tooling.Failure
+import org.gradle.tooling.GradleConnectionException
+import org.junit.Ignore
+
+@ToolingApiVersion('>=8.12')
+@TargetGradleVersion('>=8.12')
+class BuildFailureProblemsCrossVersionSpec extends ToolingApiSpecification {
+
+    def setup() {
+        settingsFile << """
+            rootProject.name = 'root'
+        """
+        buildFile << """
+            plugins {
+                id 'java-library'
+            }
+        """
+
+    }
+
+    @TargetGradleVersion('<8.12')
+    def "clients won't receive problems associated to build failures if they are not subscribed to problems"() {
+        when:
+        withConnection { connection ->
+            connection.newBuild()
+                .forTasks('doesNotExist')
+                .withDetailedFailure()
+                .run()
+        }
+
+        then:
+        GradleConnectionException exception = thrown(GradleConnectionException)
+        exception.failures.isEmpty()
+    }
+
+
+    def "build failure contains single problem report"() {
+        when:
+        withConnection { connection ->
+            connection.newBuild()
+                .forTasks('doesNotExist')
+                .withDetailedFailure()
+                .run()
+        }
+
+        then:
+        GradleConnectionException e = thrown(GradleConnectionException)
+        e.failures.size() == 1
+        e.failures[0] instanceof Failure
+        (e.failures[0].causes[0]).problems[0].contextualLabel.contextualLabel == "Task 'doesNotExist' not found in root project 'root'."
+        (e.failures[0].causes[0]).problems[0].definition.id.displayName == 'Selection failed'
+    }
+
+    def "failure does not contains report from the previous build"() {
+        when:
+        Exception firstBuildFailure = null
+        withConnection { connection ->
+            try {
+                connection.newBuild()
+                    .forTasks('doesNotExist1')
+                    .withDetailedFailure()
+                    .run()
+
+            } catch (GradleConnectionException e) {
+                firstBuildFailure = e
+            }
+            assert firstBuildFailure != null
+            connection.newBuild()
+                .forTasks('doesNotExist2')
+                .withDetailedFailure()
+                .run()
+        }
+
+        then:
+        GradleConnectionException e = thrown(GradleConnectionException)
+        e.failures.size() == 1
+        e.failures[0] instanceof Failure
+        e.failures[0].causes[0].problems[0].contextualLabel.contextualLabel == "Task 'doesNotExist2' not found in root project 'root'."
+        e.failures[0].causes[0].problems[0].definition.id.displayName == 'Selection failed'
+    }
+
+
+    @Ignore("TODO (donat) the problems API has the exception->problem association, but the worker serializes the thrown exception separately. The serialized exception instance does not equal to the one stored in the problems service")
```

Comment:
❌Unresolved TODO

Reply:
Initially I wanted to tackle this separately, but in fact the PR contains the solution. `@Ignore` removed.